test/ext/chainrulescore.jl

```diff
@@ -0,0 +1,40 @@
+using Enzyme
+using Test
+using ChainRules
+using ChainRulesCore
+using LinearAlgebra
+using EnzymeTestUtils
+
+# utility functions
+fdiff(f, x::Number) = autodiff(Forward, f, Duplicated, Duplicated(x, one(x)))[2]
+rdiff(f, x::Number) = autodiff(Reverse, f, Active, Active(x))[1][1]
+rdiff(f, x::AbstractArray) = autodiff(Reverse, f, Active, Duplicated(x, zero(x)))[1][1]
+
+@testset "import_frule" begin
+    f1(x) = 2*x
+    ChainRulesCore.@scalar_rule f1(x)  (5.0,)
+    Enzyme.@import_frule typeof(f1) Any
+    @test fdiff(f1, 1.0) === 5.0
+
+    # specific signature    
+    f2(x) = 2*x
+    ChainRulesCore.@scalar_rule f2(x)  (5*one(x),)
+    Enzyme.@import_frule typeof(f2) Float32
+    @test fdiff(f2, 1f0) === 5f0
+    @test fdiff(f2, 1.0) === 2.0
+
+    @testset "batch duplicated" begin 
+        x = [1.0, 2.0, 0.0]        
+        Enzyme.@import_frule typeof(Base.sort)  Any
+        for Tret in (Duplicated, DuplicatedNoNeed)
+            for Tx in (Duplicated, BatchDuplicated)
+                test_forward(sort, Tret, (x, Tx))
```

Comment:
this fails for Tx == BatchDuplicated. Any suggestions?

Reply:
what's the failure?

Reply:
```julia
julia> x = [1.0, 2.0, 0.0];

julia> Enzyme.@import_frule typeof(Base.sort)  Any

julia> test_forward(Base.sort, Duplicated, (x, BatchDuplicated))
test_forward: sort with return activity Duplicated on (::Vector{Float64}, BatchDuplicated): Error During Test at /Users/carlo/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:68
  Got exception outside of a @test
  DimensionMismatch: arrays could not be broadcast to a common size; got a dimension with lengths 3 and 6
  Stacktrace:
    [1] _bcs1
      @ ./broadcast.jl:555 [inlined]
    [2] _bcs
      @ ./broadcast.jl:549 [inlined]
    [3] broadcast_shape
      @ ./broadcast.jl:543 [inlined]
    [4] combine_axes
      @ ./broadcast.jl:524 [inlined]
    [5] instantiate
      @ ./broadcast.jl:306 [inlined]
    [6] materialize
      @ ./broadcast.jl:903 [inlined]
    [7] (::FiniteDifferences.var"#85#86"{FiniteDifferences.var"#87#88"{EnzymeTestUtils.var"#fnew#28"{EnzymeTestUtils.var"#call_with_copy#38"{@NamedTuple{}}, Tuple{typeof(sort), Vector{Float64}}, Tuple{Bool, Bool}}, typeof(identity)}, Vector{Float64}, Vector{Float64}})(ε::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/zWRHl/src/grad.jl:48
    [8] newf
      @ ~/.julia/packages/StaticArrays/EHHaF/src/broadcast.jl:186 [inlined]
    [9] macro expansion
      @ ~/.julia/packages/StaticArrays/EHHaF/src/broadcast.jl:135 [inlined]
   [10] __broadcast
      @ ~/.julia/packages/StaticArrays/EHHaF/src/broadcast.jl:123 [inlined]
   [11] _broadcast
      @ ~/.julia/packages/StaticArrays/EHHaF/src/broadcast.jl:119 [inlined]
   [12] copy
      @ ~/.julia/packages/StaticArrays/EHHaF/src/broadcast.jl:60 [inlined]
   [13] materialize
      @ ./broadcast.jl:903 [inlined]
   [14] _eval_function(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#85#86"{FiniteDifferences.var"#87#88"{EnzymeTestUtils.var"#fnew#28"{EnzymeTestUtils.var"#call_with_copy#38"{@NamedTuple{}}, Tuple{typeof(sort), Vector{Float64}}, Tuple{Bool, Bool}}, typeof(identity)}, Vector{Float64}, Vector{Float64}}, x::Float64, step::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/zWRHl/src/methods.jl:249
   [15] _estimate_magnitudes(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#85#86"{FiniteDifferences.var"#87#88"{EnzymeTestUtils.var"#fnew#28"{EnzymeTestUtils.var"#call_with_copy#38"{@NamedTuple{}}, Tuple{typeof(sort), Vector{Float64}}, Tuple{Bool, Bool}}, typeof(identity)}, Vector{Float64}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/zWRHl/src/methods.jl:378
   [16] estimate_step(m::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#85#86"{FiniteDifferences.var"#87#88"{EnzymeTestUtils.var"#fnew#28"{EnzymeTestUtils.var"#call_with_copy#38"{@NamedTuple{}}, Tuple{typeof(sort), Vector{Float64}}, Tuple{Bool, Bool}}, typeof(identity)}, Vector{Float64}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/zWRHl/src/methods.jl:365
   [17] (::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}})(f::FiniteDifferences.var"#85#86"{FiniteDifferences.var"#87#88"{EnzymeTestUtils.var"#fnew#28"{EnzymeTestUtils.var"#call_with_copy#38"{@NamedTuple{}}, Tuple{typeof(sort), Vector{Float64}}, Tuple{Bool, Bool}}, typeof(identity)}, Vector{Float64}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/zWRHl/src/methods.jl:193
   [18] _jvp
      @ ~/.julia/packages/FiniteDifferences/zWRHl/src/grad.jl:48 [inlined]
   [19] jvp(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::EnzymeTestUtils.var"#fnew#28"{EnzymeTestUtils.var"#call_with_copy#38"{@NamedTuple{}}, Tuple{typeof(sort), Vector{Float64}}, Tuple{Bool, Bool}}, ::Tuple{Vector{Float64}, Tuple{Vector{Float64}, Vector{Float64}}})
      @ FiniteDifferences ~/.julia/packages/FiniteDifferences/zWRHl/src/grad.jl:60
   [20] _fd_forward(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::Function, rettype::Type, y::Vector{Float64}, activities::Tuple{Const{typeof(sort)}, BatchDuplicated{Vector{Float64}, 2}})
      @ EnzymeTestUtils ~/.julia/packages/EnzymeTestUtils/CV2ed/src/finite_difference_calls.jl:30
   [21] macro expansion
      @ ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:77 [inlined]
   [22] macro expansion
      @ ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [23] test_forward(f::Function, ret_activity::Type, args::Tuple{Vector{Float64}, UnionAll}; fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, fkwargs::@NamedTuple{}, rtol::Float64, atol::Float64, testset_name::Nothing)
      @ EnzymeTestUtils ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:70
   [24] test_forward(f::Function, ret_activity::Type, args::Tuple{Vector{Float64}, UnionAll})
      @ EnzymeTestUtils ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:53
```

Reply:
@sethaxen this looks like a bug in enzymetestutils?

Reply:
As far as I know, no. As I asked in https://github.com/EnzymeAD/Enzyme.jl/pull/1264#discussion_r1488138588, does Enzyme now allow one to mix `Duplicated` and `BatchDuplicated`? That used to cause an error. EnzymeTestUtils assumes these are not mixed and provides `are_activities_compatible` to skip cases in `TestSet`s that would mix them.

Reply:
Ah, no you need one single batch width for the whole program (e.g. all duplicated, or all batchduplicated with the same width).

However, we do automatically upgrade a duplicated/dupicatednoneed return to whatever the width of the args were, if they were batch (since no data is in the return).

It would be nice for this shorthand to work. But indeed @CarloLucibello the alternate is testing {Duplicated, DuplicatedNoNeed} ret x {Const, Duplicated} input and {BatchDuplicated, BatchDuplicatedNoNeed} ret x {Const, BatchDuplicated} input

Reply:
Irrespectively, either we should upgrade testutils to handle this case (since here enzyme actually supports this, by upgrading to batchduplicated), or we should throw a nicer error here rather than bailing out in finite differences internals.

Reply:
Ok. These 4 are fine
```julia
test_forward(Base.sort, Duplicated, (x, Duplicated))
test_forward(Base.sort, BatchDuplicated, (x, BatchDuplicated))
test_forward(Base.sort, DuplicatedNoNeed, (x, Duplicated))
test_forward(Base.sort, BatchDuplicatedNoNeed, (x, BatchDuplicated))
```
But anything involving `Const` errors
```julia
test_forward(Base.sort, DuplicatedNoNeed, (x, Const))
```
```
UndefVarError: `ChainRulesCore` not defined
  Stacktrace:
    [1] forward
      @ ~/.julia/dev/Enzyme/ext/EnzymeChainRulesCoreExt.jl:67
    [2] forward
      @ ~/.julia/dev/Enzyme/ext/EnzymeChainRulesCoreExt.jl:62 [inlined]
    [3] call_with_kwargs
      @ ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:64 [inlined]
    [4] fwddiffejulia_call_with_kwargs_10402wrap
      @ ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:0
    [5] macro expansion
      @ ~/.julia/dev/Enzyme/src/compiler.jl:5440 [inlined]
    [6] enzyme_call
      @ ~/.julia/dev/Enzyme/src/compiler.jl:5118 [inlined]
    [7] ForwardModeThunk
      @ ~/.julia/dev/Enzyme/src/compiler.jl:5003 [inlined]
    [8] autodiff
      @ ~/.julia/dev/Enzyme/src/Enzyme.jl:384 [inlined]
    [9] autodiff(::ForwardMode{FFIABI}, ::EnzymeTestUtils.var"#call_with_kwargs#39"{@NamedTuple{}}, ::Type{DuplicatedNoNeed}, ::Const{typeof(sort)}, ::Const{Vector{Float64}})
      @ Enzyme ~/.julia/dev/Enzyme/src/Enzyme.jl:287
   [10] macro expansion
      @ ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:79 [inlined]
   [11] macro expansion
      @ ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [12] test_forward(f::Function, ret_activity::Type, args::Tuple{Vector{Float64}, UnionAll}; fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, fkwargs::@NamedTuple{}, rtol::Float64, atol::Float64, testset_name::Nothing)
      @ EnzymeTestUtils ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:70
   [13] test_forward(f::Function, ret_activity::Type, args::Tuple{Vector{Float64}, UnionAll})
      @ EnzymeTestUtils ~/.julia/packages/EnzymeTestUtils/CV2ed/src/test_forward.jl:53
   [14] macro expansion
      @ ~/.julia/dev/Enzyme/test/ext/chainrulescore.jl:34 [inlined]
   [15] macro expansion
      @ ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [16] macro expansion
      @ ~/.julia/dev/Enzyme/test/ext/chainrulescore.jl:25 [inlined]
   [17] macro expansion
      @ ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [18] top-level scope
      @ ~/.julia/dev/Enzyme/test/ext/chainrulescore.jl:11
```
I get errors for 
```julia
 test_forward(Base.sort, DuplicatedNoNeed, (x, Const))
 test_forward(Base.sort, Duplicated, (x, Const))
 test_forward(Base.sort, BatchDuplicatedNoNeed, (x, Const))
 test_forward(Base.sort, BatchDuplicated, (x, Const))
```

Reply:
Huh, yeah we definitely need that one to pass. Maybe try changing 
`:($ty <: Const ? ChainRulesCore.NoTangent() : $val.dval)` into 
`:($ty <: Const ? $(ChainRulesCore.NoTangent()) : $val.dval)` to fix it? (and similar throughout).

cc @vchuravy since Julia macros are not my forte